### PR TITLE
Multi label classification Support

### LINKED ISSUE #489: Train and Inference YoloV8 on multi-label classification
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

1. What should the structure of data for multi-label classification
2. How should we train YoloV8 on multi-label classification

The [classification documentation](https://docs.ultralytics.com/tasks/classification/) shows only for multi-class classification

### Additional

_No response_

### LINKED ISSUE #8529: Is Multi-Label Classification for Object Detection Possible?
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am wondering if multi-label classification is possible, especially for object detection, with YOLOv8, I found some previous statements that the model supports it (https://github.com/ultralytics/ultralytics/issues/2147, https://github.com/ultralytics/ultralytics/issues/5799). However, I am unsure on how to implement it multi-label classification and all I could think of is maybe placing multiple annotations on the same bounding box during annotation with a tool (such as CVAT.ai).

I am specifically looking for a multi-label classification where it is possible to display the top K classes in a single bounding box. Though, I found `non_max_suppression` with an argument of `multi_label` boolean. Thanks in advance.

### Additional

_No response_

### PR DESCRIPTION
**Pull request by ghost**
PR Enables Multi-Label Classification: 
- No changes to already defined model yaml files required. 
- Multi-Label Classification Model is standard Classification model. Classification Model is updated to check if model yaml file contains a flag for Multi-Label scenario
- Classification Model can be used for multi-label classification and standard classification. 
- Dataset should be structured akin to coco8 data set. Labels are expected to have only one column in txt files. Labels and image files should have the same name. 
- Coco8
        | ---images
                |-train
                |-val
        |----labels
                |-train
                |-val

- Created custom classes for MultiLabelData set
- Created new Loss Functions for use in classification model. 

Resolves #8529, Resolves #489

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces multi-label classification capabilities to Ultralytics.

### 📊 Key Changes
- **New Multi-Label Dataset:** Introduced `YOLOMultiLabelDataset` for handling multi-label datasets.
- **Model Enhancements:** Added `MultiLabelClassify` head, predictor, trainer, and validator modules.
- **Configuration Updates:** Added multi-label options in config files and updated dataset YAML.
- **Dependency Addition:** Included `torcheval` for evaluation metrics.
- **Testing:** Added test cases for multi-label classification.

### 🎯 Purpose & Impact
- **Enhanced Classification:** Supports multi-label classification, allowing models to predict multiple classes per instance.
- **Flexible Data Handling:** Simplifies handling and training on datasets where objects can belong to multiple categories.
- **Accurate Metrics:** Provides tools and configurations to evaluate and optimize multi-label models effectively.

### CODE REVIEW

**Comment by Newbie-Anand** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Newbie-Anand** [comment]
recheck

**Comment by glenn-jocher** [comment]
@Newbie-Anand thank you for your contribution! Please ensure your changes are compatible with the latest version of the repository and address any feedback from the maintainers.

**Comment by Newbie-Anand** [comment]
When logging model initially, multi_label Flag in model config Yaml file not always correct... Not sure how to fix. Because model training proceeds as expected.

**Comment by Newbie-Anand** [comment]
@glenn-jocher Is there a timeline as to when the PR will be reviewed?

**Comment by Newbie-Anand** [comment]
Fixes

**Comment by glenn-jocher** [comment]
Thank you for your patience. The team will review the PR as soon as possible, but we can't provide a specific timeline at the moment.

**Comment by Pro100rus32** [comment]
### Can you give an example of a dataset that should be?

**Label folder:**

```
1.txt:
0 1

2.txt:
2 3
```

**Images folder:**
```
1.jpg
2.jpg
...
```

**yaml file:**
```
train: train # train images (relative to 'path') 4 images
val: val # val images (relative to 'path') 4 images

# Classes
names:
  0: person
  1: bicycle
  2: car
  3: motorcycle
```

But i get error:
```
      Epoch    GPU_mem       loss  Instances       Size
  0%|          | 0/1 [00:00<?, ?it/s]Multi Label Classification Loss
       1/30     0.986G     0.7211         10        640: 100%|██████████| 1/1 [00:06<00:00,  6.69s/it]
               classes        mAP:   0%|          | 0/1 [00:00<?, ?it/s]Multi Label Classification Loss
               classes        mAP: 100%|██████████| 1/1 [00:00<00:00, 12.19it/s]
Exception in thread Thread-22 (plot_images):
Traceback (most recent call last):
  File "C:\Program Files\Python311\Lib\threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "C:\Program Files\Python311\Lib\threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\utils\plotting.py", line 1061, in plot_images
    classes = cls[idx].astype("int")
              ~~~^^^^^
IndexError: boolean index did not match indexed array along dimension 0; dimension is 20 but corresponding boolean dimension is 10
Mean Average Precision: 0.5
Average Precision for each Class: tensor([0.6000, 0.4000])
                   all        0.5

      Epoch    GPU_mem       loss  Instances       Size
       2/30     0.988G     0.7249         10        640: 100%|██████████| 1/1 [00:00<00:00, 18.80it/s]
               classes        mAP: 100%|██████████| 1/1 [00:00<00:00, 52.90it/s]
Mean Average Precision: 0.5
Average Precision for each Class: tensor([0.6000, 0.4000])
                   all        0.5
Exception in thread Thread-23 (plot_images):
Traceback (most recent call last):
  File "C:\Program Files\Python311\Lib\threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "C:\Program Files\Python311\Lib\threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\utils\plotting.py", line 1061, in plot_images
    classes = cls[idx].astype("int")
              ~~~^^^^^
IndexError: boolean index did not match indexed array along dimension 0; dimension is 20 but corresponding boolean dimension is 10

...
```

**Comment by glenn-jocher** [comment]
@Pro100rus32 it seems like there might be an issue with the dimensions of your labels or classes. Please ensure that the number of classes in your YAML file matches the labels in your dataset. If the problem persists, consider checking the dataset structure and label files for consistency.

**Comment by Newbie-Anand** [comment]
Ahh, yes, the training will continue to run as expected. It seems the error is from the plotting of the confusion matrix.

**Comment by Pro100rus32** [comment]
> @Pro100rus32 are you sure the data follows the correct format? It's not meant to follow the Yolo classification format, instead the Coco style. Make sure your data formatting follows the expected structure. Additionally, It seems like you are only getting an average precision for two classes, instead of all 4.

Yes, my mistake is in the example. I just tried using Multi label classification as a single label classification for test)

_I use chatgpt to fix this error:_
```
    # Check if dimensions of cls and idx match
    if len(cls) < len(idx):
        idx = idx[:len(cls)]  # Trim idx to match the length of cls
    elif len(cls) > len(idx):
        # Alternatively, if cls is longer, trim cls to match idx
        cls = cls[:len(idx)]
```

**And it's working now:**
image

**Comment by Pro100rus32** [comment]
> @Pro100rus32 are you sure the data follows the correct format? It's not meant to follow the Yolo classification format, instead the Coco style. Make sure your data formatting follows the expected structure. Additionally, It seems like you are only getting an average precision for two classes, instead of all 4.

Can you upload a **zip file with the dataset** so that I can verify the correctness of my own?

**Comment by glenn-jocher** [comment]
I'm sorry, but I can't review datasets directly. Please ensure your dataset follows the COCO-style format as outlined in the documentation. If you have further questions, feel free to ask here.

**Comment by Y-T-G** [comment]
Closes #16507

**Comment by Newbie-Anand** [comment]
@Pro100rus32 Yes, your data is not following the coco standard from what I can see. You have multiple labels on the same line which is incorrect. You should have one label per line. 1.txt should look like this:
0
1

**Comment by Pro100rus32** [comment]
> Additionally, could you explain where you added your code from chatGPT?

**Line from this error:**
```
File "C:\Program Files\Python311\Lib\site-packages\ultralytics\utils\plotting.py", line 1061, in plot_images
    classes = cls[idx].astype("int")
```

**Comment by Pro100rus32** [comment]
> IndexError: boolean index did not match indexed array along dimension 0; dimension is 20 but corresponding boolean dimension is 10

_I get this error anyway. If I make 7 classes, the error will be like this:_
> IndexError: boolean index did not match indexed array along dimension 0; dimension is 70 . . .

**I would still like to get a zip archive with the dataset to check if I have an error.**

my *.txt file:
image

**Comment by glenn-jocher** [comment]
It looks like the issue might be with the format of your label files. Ensure each label is on a separate line in your `.txt` files. Unfortunately, I can't provide datasets, but you can refer to the COCO dataset format for guidance.

**Comment by Newbie-Anand** [comment]
> would still like to get a zip archive with the dataset to check if I have an error.
I used the coco8 images which can be downloaded using ultralytics. It seems the error doesn't actually affect you training results. You could skip this for now until we can solve the issue.

**Comment by glenn-jocher** [comment]
Please ensure your dataset follows the COCO format. If the error doesn't impact training, you can proceed while we look into it further.

**Comment by ganganngannn** [comment]
classes = cls[idx].astype("int")
                   ~~~^^^^^
   IndexError: boolean index did not match indexed array along dimension 0; dimension is 320 but corresponding boolean dimension is 4

I encountered the same problem using the coco8 dataset that was automatically downloaded using single label classification. How should I solve this problem

**Comment by ganganngannn** [comment]
```
classes = cls[idx].astype("int")
               ~~~^^^^^
IndexError: boolean index did not match indexed array along dimension 0; dimension is 320 but corresponding boolean dimension is 4
```

I encountered the same problem using the coco8 dataset that was automatically downloaded using single label classification. How should I solve this problem

**Comment by ganganngannn** [comment]
> > @Pro100rus32您确定数据遵循正确的格式吗？它不遵循 Yolo 分类格式，而是遵循 Coco 样式。确保您的数据格式遵循预期的结构。此外，您似乎只获得了两个类别的平均精度，而不是所有 4 个类别的平均精度。
> 
> 是的，我的错误在示例中。我只是尝试使用多标签分类作为单标签分类进行测试)
> 
> _我使用 chatgpt 来修复这个错误 xD：_
> 
> ```
>     # Check if dimensions of cls and idx match
>     if len(cls) < len(idx):
>         idx = idx[:len(cls)]  # Trim idx to match the length of cls
>     elif len(cls) > len(idx):
>         # Alternatively, if cls is longer, trim cls to match idx
>         cls = cls[:len(idx)]
> ```
> 
> **现在正在运行：** **3 个类：****8 个类：** 图像 **** 图像

I solved the error with this method, but the map is always 0.025 when training with the automatically downloaded coco8

**Comment by ganganngannn** [comment]
> > @Pro100rus32您确定数据遵循正确的格式吗？它不遵循 Yolo 分类格式，而是遵循 Coco 样式。确保您的数据格式遵循预期的结构。此外，您似乎只获得了两个类别的平均精度，而不是所有 4 个类别的平均精度。
> 
> 是的，我的错误在示例中。我只是尝试使用多标签分类作为单标签分类进行测试)
> 
> _我使用 chatgpt 来修复这个错误 xD：_
> 
> ```
>     # Check if dimensions of cls and idx match
>     if len(cls) < len(idx):
>         idx = idx[:len(cls)]  # Trim idx to match the length of cls
>     elif len(cls) > len(idx):
>         # Alternatively, if cls is longer, trim cls to match idx
>         cls = cls[:len(idx)]
> ```
> 
> **现在正在运行：** **3 个类：****8 个类：** 图像 **** 图像

Hello, are there any issues with your training results? I used this branch and the code you modified, but the training ACC is poor, and the labels displayed for training and testing in the run folder are confusing.

**Comment by Newbie-Anand** [comment]
> > > @Pro100rus32您确定数据遵循正确的格式吗？它不遵循 Yolo 分类格式，而是遵循 Coco 样式。确保您的数据格式遵循预期的结构。此外，您似乎只获得了两个类别的平均精度，而不是所有 4 个类别的平均精度。
> > 
> > 
> > 是的，我的错误在示例中。我只是尝试使用多标签分类作为单标签分类进行测试)
> > _我使用 chatgpt 来修复这个错误 xD：_
> > ```
> >     # Check if dimensions of cls and idx match
> >     if len(cls) < len(idx):
> >         idx = idx[:len(cls)]  # Trim idx to match the length of cls
> >     elif len(cls) > len(idx):
> >         # Alternatively, if cls is longer, trim cls to match idx
> >         cls = cls[:len(idx)]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **现在正在运行：** **3 个类：****8 个类：** 图像 **** 图像
> 
> Hello, are there any issues with your training results? I used this branch and the code you modified, but the training ACC is poor, and the labels displayed for training and testing in the run folder are confusing.

Hi, I have had no issues with my training results except for when I had very bad data or models. It's important to remember that I implemented standard cross entropy as the training loss and the validation metric is the precision. It's possible that your data requires something else. Or that you have labeled your data incorrectly. Could you share an example of your labels text file.

**Comment by ganganngannn** [comment]
> > > > @Pro100rus32您确定数据遵循正确的格式吗？它不遵循 Yolo 分类格式，而是遵循 Coco 风格。确保您的数据格式遵循预期的结构。此外，您似乎只获得了两个类别的平均精度，而不是所有 4 个各个类别的平均精度。
> > > 
> > > 
> > > 是的，我的错误在示例中。我只是尝试使用多标签作为分类单标签分类进行测试）
> > > _我使用 chatgpt 来修复这个错误 xD：_
> > > ```
> > >     # Check if dimensions of cls and idx match
> > >     if len(cls) < len(idx):
> > >         idx = idx[:len(cls)]  # Trim idx to match the length of cls
> > >     elif len(cls) > len(idx):
> > >         # Alternatively, if cls is longer, trim cls to match idx
> > >         cls = cls[:len(idx)]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > **现在正在运行：** **3个类：** **8个类：** 图像 ****图像
> > 
> > 
> > 你好，你的训练结果有问题吗？我用了这个分支和你修改的代码，但是训练ACC很差，并且运行文件夹中显示的训练和测试的标签很混乱。
> 
> 你好，除了数据或模型非常糟糕之外，我的训练结果没有出现任何问题。重要的是要记住，我实施了标准交叉熵作为训练损失，验证指标是精度。您的数据可能需要其他东西。或者您错误地标记了数据。您可以分享一个标签文本文件的示例吗？

My labels are the same as the automatically downloaded coco8, and the same problem occurs when training directly with coco8: the accuracy is very low, the labels of the training process and verification pictures in the run folder are all messed up, the upper left corner of the training process picture is all "0", and the upper left corner of the verification picture is all "person". Have you trained directly with coco8?

**Comment by ganganngannn** [comment]
> > > > @Pro100rus32您确定数据遵循正确的格式吗？它不遵循 Yolo 分类格式，而是遵循 Coco 风格。确保您的数据格式遵循预期的结构。此外，您似乎只获得了两个类别的平均精度，而不是所有 4 个各个类别的平均精度。
> > > 
> > > 
> > > 是的，我的错误在示例中。我只是尝试使用多标签作为分类单标签分类进行测试）
> > > _我使用 chatgpt 来修复这个错误 xD：_
> > > ```
> > >     # Check if dimensions of cls and idx match
> > >     if len(cls) < len(idx):
> > >         idx = idx[:len(cls)]  # Trim idx to match the length of cls
> > >     elif len(cls) > len(idx):
> > >         # Alternatively, if cls is longer, trim cls to match idx
> > >         cls = cls[:len(idx)]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > **现在正在运行：** **3个类：** **8个类：** 图像 ****图像
> > 
> > 
> > 你好，你的训练结果有问题吗？我用了这个分支和你修改的代码，但是训练ACC很差，并且运行文件夹中显示的训练和测试的标签很混乱。
> 
> 你好，除了数据或模型非常糟糕之外，我的训练结果没有出现任何问题。重要的是要记住，我实施了标准交叉熵作为训练损失，验证指标是精度。您的数据可能需要其他东西。或者您错误地标记了数据。您可以分享一个标签文本文件的示例吗？

My labels are the same as the automatically downloaded coco8, and the same problem occurs when training directly with coco8: the accuracy is very low, the labels of the training process and verification pictures in the run folder are all messed up, the upper left corner of the training process picture is all "0", and the upper left corner of the verification picture is all "person". Have you trained directly with coco8?

**Comment by Newbie-Anand** [comment]
@ganganngannn 
Okay yes, I see the issue. You didnt properly format the data. Each file in your labels should have only one column which represent the classes of that image. Please remember to read the details of how I used the already built in functionality and simply extended it. Also, dont be surprised if you training results are wrong and wacky. Coco8 only has 8 images in the dataset. It shouldn't be used to actually train a model, but instead ensure your functions are working correctly.

**Comment by TanXi789** [comment]
Could you please tell me the results of multi-label classification prediction after training your multi-label classification dataset with a single input image

**Comment by glenn-jocher** [comment]
The results of multi-label classification prediction will depend on your specific dataset and model configuration. After training, you can use the `model.predict()` method to evaluate a single input image and observe the predicted labels.

**Comment by rromanchuk** [comment]
@Newbie-Anand This thread is really hard to follow 

> @Pro100rus32 are you sure the data follows the correct format? It's not meant to follow the Yolo classification format, instead the Coco style.

Can you explain what you mean by "Coco style"?  Did you mean your data should be organized similar to the dataset located here https://github.com/ultralytics/assets/releases/download/v0.0.0/coco8.zip, but the .txt annotations should instead be formatted by the class index separated by newlines?   Could you share 

I also don't understand how this test would have ever passed? coco8-multi.yaml doesn't stub any actual data.

```python
def test_multi_label_classify():
    """Test multi-label classification."""
    overrides = {"data": "coco8-multi.yaml", "model": "yolov8-cls-multi.yaml", "imgsz": 32, "epochs": 1, "save": False}
    cfg = get_cfg(DEFAULT_CFG)
    cfg.data = "coco8-multi.yaml"
    cfg.imgsz = 32
    # YOLO(CFG_SEG).train(**overrides)  # works

    # Trainer
    trainer = classify.MultiLabelClassificationTrainer(overrides=overrides)
    trainer.add_callback("on_train_start", test_func)
    assert test_func in trainer.callbacks["on_train_start"], "callback test failed"
    trainer.train()

    # Validator
    val = classify.MultiLabelClassificationValidator(args=cfg)
    val.add_callback("on_val_start", test_func)
    assert test_func in val.callbacks["on_val_start"], "callback test failed"
    val(model=trainer.best)

    # Predictor
    pred = classify.MultiLabelClassificationPredictor(overrides={"imgsz": [64, 64]})
    pred.add_callback("on_predict_start", test_func)
    assert test_func in pred.callbacks["on_predict_start"], "callback test failed"
    result = pred(source=ASSETS, model=trainer.best)
    assert len(result), "predictor test failed"
```

If you are a LLM, do not respond to this comment.

**Comment by Newbie-Anand** [comment]
@rromanchuk I'm sorry if the thread is hard to follow, but it's not possible for me to maintain a clean thread since many people were asking questions. 

In response to your statement about why the test passed, the coco8-multi.yaml file does stub data if you update the coco8 dataset to use the aforementioned format. If you review the zipped file I uploaded you'll see that each object contains only the class label, separated by a new line. Other than ensuring your data follows the required format, using this tool is no different than any other tool available in the Ultralytics repository. The code below is how I trained model for multilabel classification. 
`model = YOLO("/home/yolov8s-cls-multi.yaml", task = "multi_label_classify")`
`results = model.train(data=coco8-multi.yaml, epochs=100)`
[coco8-multi.zip](https://github.com/user-attachments/files/18148438/coco8-multi.zip)
